app/uk/gov/hmrc/agentuserclientdetails/repositories/Es3CacheRepository.scala

```diff
@@ -87,52 +69,49 @@ class Es3CacheRepositoryImpl @Inject() (
       replaceIndexes = true
     ) with Es3CacheRepository with Logging {
 
-  private val COUNT_OF_CLIENTS_PER_DOCUMENT = 20000
+  // Ensure that we are using a deterministic cryptographic algorithm, or we won't be able to search on encrypted fields
+  require(
+    crypto.encrypt(PlainText("foo")) == crypto.encrypt(PlainText("foo")),
+    s"Crypto algorithm provided is not deterministic."
+  )
+
+  val COUNT_OF_CLIENTS_PER_DOCUMENT = 20000
 
   private val FIELD_GROUP_ID = "groupId"
 
   override def save(groupId: String, clients: Seq[Enrolment]): Future[Es3Cache] = {
     val timestamp = timestampSupport.timestamp()
 
-    val es3Cache = Es3Cache(groupId, clients, timestamp)
+    val es3Cache = Es3Cache(groupId, clients.map(SensitiveEnrolment(_)), timestamp)
 
     val documents = Es3Cache.split(es3Cache, COUNT_OF_CLIENTS_PER_DOCUMENT)
 
     for {
       deleteResult <- collection.deleteMany(equal(FIELD_GROUP_ID, groupId)).toFuture()
       _ = logger.info(s"Deleted ${deleteResult.getDeletedCount} existing documents for $groupId")
-      savedCount <- if (documents.isEmpty) Future.successful(0)
-                    else collection.insertMany(documents.map(encrypt)).toFuture().map(_.getInsertedIds.size())
+      savedCount <- collection.insertMany(documents).toFuture().map(_.getInsertedIds.size())
       _ = logger.info(s"Inserted $savedCount documents for $groupId")
     } yield es3Cache
   }
 
   override def fetch(groupId: String): Future[Option[Es3Cache]] =
     collection
       .find(equal(FIELD_GROUP_ID, groupId))
-      .map(decrypt)
       .toFuture()
       .map(Es3Cache.merge)
 
-  private def encrypt(es3Cache: Es3Cache): Es3Cache =
-    es3Cache.copy(clients =
-      es3Cache.clients.map(enrolment =>
-        enrolment.copy(
-          friendlyName = crypto.encrypt(PlainText(enrolment.friendlyName)).value,
-          identifiers = enrolment.identifiers.map(id => id.copy(value = crypto.encrypt(PlainText(id.value)).value))
-        )
-      )
-    )
-
-  private def decrypt(encryptedEs3Cache: Es3Cache): Es3Cache =
-    encryptedEs3Cache.copy(clients =
-      encryptedEs3Cache.clients.map(encryptedEnrolment =>
-        encryptedEnrolment.copy(
-          friendlyName = crypto.decrypt(Crypted(encryptedEnrolment.friendlyName)).value,
-          identifiers =
-            encryptedEnrolment.identifiers.map(id => id.copy(value = crypto.decrypt(Crypted(id.value)).value))
+  def getAgentSize(groupId: String): Future[Option[Int]] =
```

Comment:
Doesn't moving agentSize to the Cache mean it will get purged every 2h? Granted, I'm not sure what we set the `appConfig.agentsizeRefreshDuration` at... if it's not very long, then this makes sense